packetbeat/protos/amqp/amqp.go

```diff
@@ -0,0 +1,883 @@
+	amqp.MethodMap[CONNECTION][CONNECTIONSTART] = connectionStartMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONSTARTOK] = connectionStartOkMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONTUNE] = connectionTuneMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONTUNEOK] = connectionTuneOkMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONOPEN] = connectionOpenMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONOPENOK] = okMethod
+	amqp.MethodMap[CHANNEL][CHANNELOPEN] = channelOpenMethod
+	amqp.MethodMap[CHANNEL][CHANNELOPENOK] = okMethod
+	amqp.MethodMap[CHANNEL][CHANNELFLOW] = channelFlowMethod
+	amqp.MethodMap[CHANNEL][CHANNELFLOWOK] = channelFlowOkMethod
+	amqp.MethodMap[BASIC][BASICQOS] = basicQosMethod
+	amqp.MethodMap[BASIC][BASICQOSOK] = okMethod
+}
+
+func (amqp *Amqp) ConnectionTimeout() time.Duration {
+	return amqp.transactionTimeout
+}
+
+func (amqp *Amqp) Parse(pkt *protos.Packet, tcptuple *common.TcpTuple,
+	dir uint8, private protos.ProtocolData) protos.ProtocolData {
+
+	defer logp.Recover("ParseAmqp exception")
+	logp.Debug("amqpdetailed", "Parse method triggered")
+
+	priv := amqpPrivateData{}
+	if private != nil {
+		var ok bool
+		priv, ok = private.(amqpPrivateData)
+		if !ok {
+			priv = amqpPrivateData{}
+		}
+	}
+
+	if priv.Data[dir] == nil {
+		priv.Data[dir] = &AmqpStream{
+			tcptuple: tcptuple,
+			data:     pkt.Payload,
+			message:  &AmqpMessage{Ts: pkt.Ts},
+		}
+	} else {
+		// concatenate databytes
+		priv.Data[dir].data = append(priv.Data[dir].data, pkt.Payload...)
+		if len(priv.Data[dir].data) > tcp.TCP_MAX_DATA_IN_STREAM {
+			logp.Debug("amqp", "Stream data too large, dropping TCP stream")
+			priv.Data[dir] = nil
+			return priv
+		}
+	}
+
+	stream := priv.Data[dir]
+
+	for len(stream.data) > 0 {
+		if stream.message == nil {
+			stream.message = &AmqpMessage{Ts: pkt.Ts}
+		}
+
+		ok, complete := amqp.amqpMessageParser(priv.Data[dir])
+		if !ok {
+			// drop this tcp stream. Will retry parsing with the next
+			// segment in it
+			priv.Data[dir] = nil
+			return priv
+		}
+		if complete {
+			amqp.handleAmqp(stream.message, tcptuple, dir)
+		} else {
+			// wait for more data
+			break
+		}
+	}
+	return priv
+}
+
+func (amqp *Amqp) GapInStream(tcptuple *common.TcpTuple, dir uint8,
+	nbytes int, private protos.ProtocolData) (priv protos.ProtocolData, drop bool) {
+	//TODO
+	return private, true
+}
+
+func (amqp *Amqp) ReceivedFin(tcptuple *common.TcpTuple, dir uint8,
+	private protos.ProtocolData) protos.ProtocolData {
+	//TODO
+	return private
+}
+
+func (amqp *Amqp) amqpMessageParser(s *AmqpStream) (ok bool, complete bool) {
+	for s.parseOffset < len(s.data) {
+
+		if len(s.data[s.parseOffset:]) < 8 {
+			logp.Warn("AMQP message smaller than a frame, discarding it")
+			return false, false
+		}
+
+		yes, version := isProtocolHeader(s.data[s.parseOffset:])
+		if yes {
+			logp.Debug("amqp", "Client header detected, version %d.%d.%d",
+				version[0], version[1], version[2])
+			s.parseOffset += 8
+			return true, false
```

Comment:
Ok so to clarify this, when I set ok = false, it means that the data received is actually wrong. Not incomplete, but _wrong_, not satisfying the specs of the protocol. I really would like to clarify this because those lines are among the first I wrote for this pull request and maybe there were some things I did not get right at the time. So if I set ok=false, the data will be dropped and the parser will take the next input of data, but if I set ok=true, the parser will just wait for more data to complete what we have. Is this right ?